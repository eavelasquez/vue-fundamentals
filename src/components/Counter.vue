<template>
  <h2>{{ customTitle }}</h2>
  <p data-testid="counter">{{ counter }}</p>
  <p>{{ counter }} <sup>2</sup> = {{ squaredCounter }}</p>
  <!-- <p>{{ counter }} <sup>2</sup> = {{ calculateSquaredCounter() }}</p> -->

  <div class="buttons-container">
    <button @click="decrement">-</button>
    <button @click="increment">+</button>
    <button @click="reset">&#8634;</button>
  </div>
</template>

<script>
export default {
  name: 'Counter',
  props: {
    title: String,
    initialCounter: {
      type: Number,
      default: 0,
      // required: true,
      // Custom validator function
      validator(value) {
        return value >= 0
      },
    },
  },
  data() {
    return {
      counter: this.initialCounter,
    }
  },
  methods: {
    // calculateSquaredCounter() {
    //   console.log('methods: calculateSquaredCounter()')
    //   return this.counter * this.counter
    // },
    decrement() {
      this.counter -= 1
    },
    increment() {
      this.counter += 1
    },
    reset() {
      this.counter = this.initialCounter
    },
  },
  computed: {
    // a computed getter
    squaredCounter() {
      console.log('computed: squaredCounter')
      // `this` points to the vm instance
      return this.counter * this.counter
    },
    customTitle() {
      return this.title || 'Current Counter'
    },
  },
}
</script>

<style scoped>
p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}
</style>