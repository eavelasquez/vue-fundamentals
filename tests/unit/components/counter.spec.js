import { shallowMount } from '@vue/test-utils'
import Counter from '@/components/Counter.vue'

describe('Counter.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(Counter)
  })

  test('should match snapshot', () => {
    const wrapper = shallowMount(Counter)

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('h2 should be the default value', () => {
    expect(wrapper.find('h2').exists()).toBeTruthy()

    const h2 = wrapper.find('h2')

    expect(h2.text()).toBe('Current Counter')
  })

  test('the default initial counter should be 7 on p', () => {
    const p = wrapper.find('[data-testid="counter"]')

    expect(p.text()).toBe('0')
  })

  test('should decrement counter when button is clicked', () => {
    const button = wrapper.find('button')

    button.trigger('click')

    expect(wrapper.vm.counter).toBe(-1);
  })

  test('should increment counter when button is clicked', () => {
    const button = wrapper.findAll('button')[1]

    button.trigger('click')
    button.trigger('click')

    expect(wrapper.vm.counter).toBe(2);
  })

  test('should reset counter when button is clicked', async () => {
    const button = wrapper.findAll('button')[2]

    await button.trigger('click')
    const p = wrapper.find('[data-testid="counter"]')

    expect(p.text()).toBe('0')
  })

  test('should set the default initial counter', () => {
    const { initialCounter } = wrapper.props()

    const p = wrapper.find('[data-testid="counter"]')

    expect(+(p.text())).toBe(initialCounter)
  })

  test('should show the prop title', () => {
    const title = 'Incoming'
    const wrapper = shallowMount(Counter, { props: { title } })
    const h2 = wrapper.find('h2')

    expect(h2.text()).toBe(title)
  })
})