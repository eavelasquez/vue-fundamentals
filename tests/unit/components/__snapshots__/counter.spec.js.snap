// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter.vue should match snapshot 1`] = `
<h2>Current Counter</h2>
<p data-testid="counter">0</p>
<p>0 <sup>2</sup> = 0</p>
<!-- <p>{{ counter }} <sup>2</sup> = {{ calculateSquaredCounter() }}</p> -->
<div class="buttons-container"><button>-</button><button>+</button><button>â†º</button></div>
`;
